### ISSUE TITLE
Listbox - focus does not move after keypress (example for Down Arrow)

### ISSUE THREAD

**Issue by dotherightthing** [w3c/aria-practices#1573]
## Problem

The examples do not match the written specification. This raises the question of whether the specification is correct.

## Example: Down Arrow

> Down Arrow: Moves focus to the next option. Optionally, in a single-select listbox, selection may also move with focus.
>
> Source: <https://www.w3.org/TR/wai-aria-practices/#listbox_kbd_interaction>

Pressing the down arrow key moves the programmatic selection to the next option (`aria-selected="true"`).

The newly 'focussed' option gains a light blue background (`class="focused"`).

But the programmatic *focus* remains on the listbox (blue focus outline, applied by the browser via the `:focus` pseudo-class).

This issue also affects the initial focus after tabbing to the element. In order to keep this report concise, I've only included an example of the Down Arrow behaviour.

## Reproducible tests

Run these tests in Chrome 86.0.4240.80 (macOS 10.15.7).

### Test 1

1. Navigate to <https://www.w3.org/TR/wai-aria-practices/examples/listbox/listbox-collapsible.html>
2. Setup tests (see heading below)
3. Tab to the Example listbox
4. Press ENTER key
5. Press Down Arrow key
6. Observe test results

### Test 2

1. Navigate to <https://www.w3.org/TR/wai-aria-practices/examples/listbox/listbox-scrollable.html>
2. Setup tests (see heading below)
3. Tab to the Example listbox
4. Press Down Arrow key
5. Observe test results

### Setup visual focus test

In Chrome Dev Tools, add a New Style Rule:

```
*:focus {
  outline: 2px dashed green;
}
```

### Setup programmatic focus test

In Chrome Dev Tools, paste the following into the console:

```
// for each listbox on the page
document.querySelectorAll('[role="listbox"]').forEach( function($listbox) {

    // after the down arrow is pressed, wait 2 seconds then test which element has focus
    // expected: second (or next) option has focus
    // actual: listbox has focus

    $listbox.addEventListener('keydown', function (e) {
       if (e.key === 'ArrowDown') {
            setTimeout( function () {
                console.log(document.activeElement);
            }, 2000);
        }
    });
});
```

**Comment by dotherightthing** [comment]
Upon further investigation I can see that the two examples do actually move the focus, by using `aria-activedescendant`:

> The listbox role supports the aria-activedescendant property, which provides an alternative to moving DOM focus among option elements when implementing keyboard navigation.
>
> Source: <https://www.w3.org/TR/wai-aria-practices/#Listbox>

**Comment by mcking65** [comment]
@dotherightthing 

You are right. This can be quite confusing to people who are reading the examples and are not familiar with focus management via aria-activedescendant. Like the ARIA spec, the APG uses the word "focus" without always specifying if the focus is document.activeElement or visual focus. In practice, it would be very challenging to always modify the word focus to be that specific. 

In attempt to help mitigate this problem, on some example pages, we have some text above the keyboard support table that says that focus refers to the element referenced by activedescendant. That is an attempt to help clarify. For instance, have a look at the [Editable Combobox With Both List and Inline Autocomplete Example](https://w3c.github.io/aria-practices/examples/combobox/combobox-autocomplete-both.html). Under the "Keyboard Support" heading, above the table for the "Listbox Popup", we have the following:

>NOTE: When visual focus is in the listbox, DOM focus remains on the textbox and the value of aria-activedescendant on the textbox is set to a value that refers to the listbox option that is visually indicated as focused. Where the following descriptions of keyboard commands mention focus, they are referring to the visual focus indicator. For more information about this focus management technique, see Using aria-activedescendant to Manage Focus.

If we would have included this note on the listbox example pages, would that have helped you?

**Comment by dotherightthing** [comment]
@mcking65

Thanks, the meaning is clear there. I think it would be better to break it into shorter sentences, for the benefit of developers with reading issues.

However, I don't really understand the benefit of aria-activedescendant - it would be interesting to know what problem that was trying to solve.

The Authoring Practices state that:

> If a component container has an ARIA role that supports the aria-activedescendant property, it is not necessary to manipulate the tabindex attribute and move DOM focus among focusable elements within the container.

Some problems I see are:

* You still need to "move the {visual} focus among focusable elements within the container". Where previously you got that for free by using basic HTML (`tabindex="-1"` on all options to allow programmatic focus), it's now something extra that you need to add in via a roaming CSS class.
* If an ARIA role doesn't support the aria-activedescendant property, you potentially need to use a different method
* Visually faking the focus by toggling a class seems to defy the semantic convention (of `tabindex` + `:focus`), which will still be useful for non-ARIA use cases such as section skip links.
* QA testers can't visually check that the ARIA has been integrated correctly, because a developer could potentially implement the CSS class but not update the value of `aria-activedescendant`.
* There's potential confusion when two elements are simultaneously presented as visually in focus (one being in `:focus`, the other in `.focus`)
* Reliance on a new property breaks support in legacy AT (I'll log my issue with the interoperability disclaimer separately!)

**Comment by carmacleod** [comment]
Hi, @dotherightthing !

Please note that `aria-activedescendant` isn't new. It actually predates the `"-1"` value for `tabindex` (which was not a thing in HTML4 when ARIA was first released). So the problem it was originally trying to solve was that there was no way to manage focus of child elements in a parent container using only positive values for `tabindex`.

It is supported on all container roles that need to manage focus of child elements.
As far as I know, even legacy AT have [support for `aria-activedescendant`](https://a11ysupport.io/tech/aria/aria-activedescendant_attribute), with the possible exception of Narrator on legacy Edge.

That said, you are correct that using `aria-activedescendant` for focus management can potentially be more difficult and error-prone than using `tabindex` in today's modern browsers. Another problem I've seen is that browsers manage scrolling for you when you move DOM focus to an element with a `tabindex`, whereas that's left up to the developer when using `aria-activedescendant`.

But then again, maybe the developer wants to control the scrolling themselves, and maybe there are use cases where a component is easier to build with `aria-activedescendant` so that the developer does not have to figure out how to suppress the free scrolling behavior because it's not quite what the component needed? I'm not sure. It may be that there are pros and cons to each method.

Perhaps the APG needs to review all uses of the word "focus", and clarify whether DOM focus or `aria-activedescendant` focus (or either) is meant in each case. Perhaps the guide needs to point more often to the [section on Managing Focus in the ARIA spec](https://w3c.github.io/aria/#managingfocus).
